arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp

```diff
@@ -101,79 +108,101 @@ auto DocumentStateTransactionHandler::getTrx(TransactionId tid)
+          if constexpr (FinishesUserTransaction<T>) {
+            auto res = Result{};
+            ADB_PROD_ASSERT(trx != nullptr);
+            if constexpr (std::is_same_v<T, ReplicatedOperation::Commit>) {
+              res = trx->commit();
+            } else if constexpr (std::is_same_v<T,
+                                                ReplicatedOperation::Abort>) {
+              res = trx->abort();
+            }
+            removeTransaction(op.tid);
+            return res;
+          } else if constexpr (std::is_same_v<
+                                   T,
+                                   ReplicatedOperation::IntermediateCommit>) {
+            ADB_PROD_ASSERT(trx != nullptr);
+            return trx->intermediateCommit();
+          } else if constexpr (ModifiesUserTransaction<T>) {
+            if (trx == nullptr) {
+              auto accessType = std::is_same_v<T, ReplicatedOperation::Truncate>
+                                    ? AccessMode::Type::EXCLUSIVE
+                                    : AccessMode::Type::WRITE;
+              trx = _factory->createTransaction(*_vocbase, op.tid, op.shard,
+                                                accessType);
+              setTrx(op.tid, trx);
+            }
+            auto opRes = trx->apply(operation);
+            auto res = opRes.fail()
+                           ? opRes.result
+                           : makeResultFromOperationResult(opRes, op.tid);
+            if (res.fail() && shouldIgnoreError(opRes)) {
+              LOG_CTX("0da00", INFO, _logContext)
+                  << "Result " << res << " ignored while applying transaction "
+                  << op.tid;
+              LOG_CTX("7eecc", DEBUG, _logContext)
+                  << "Operation failure ignored for: " << operation;
+              res = Result{};
+            }
+            return res;
+          } else {
+            static_assert(
+                always_false_v<T>,
+                "Unhandled operation type. This should never happen.");
+          }
+        } else if constexpr (std::is_same_v<
+                                 T, ReplicatedOperation::AbortAllOngoingTrx>) {
+          _transactions.clear();
           return Result{};
+        } else if constexpr (std::is_same_v<T,
+                                            ReplicatedOperation::CreateShard>) {
+          auto shardRes = _shardHandler->ensureShard(op.shard, op.collection,
+                                                     op.properties);
+          if (shardRes.fail()) {
+            return shardRes.result();
+          }
+          if (!shardRes.get()) {
+            LOG_CTX("c98e9", INFO, _logContext)
+                << "Shard " << op.shard << " already exists";
+          }
+          return Result{};
+        } else if constexpr (std::is_same_v<T,
+                                            ReplicatedOperation::DropShard>) {
+          auto transactions = getTransactionsForShard(op.shard);
```

Comment:
Please wrap this 2 lines into ifdef MAINTAINER_MODE

The transactions are only used within the Assert, so we can help the compiler that this is actually only required in Maintainer mode, and not required in release. (We probably end up with "unused" compiler warning here anyhow.

Reply:
Good catch!